admittance_controller/src/admittance_controller.cpp

```diff
@@ -0,0 +1,505 @@
+{
+  // update input reference from ros subscriber message
+  if (!admittance_)
+  {
+    return controller_interface::return_type::ERROR;
+  }
+
+  joint_command_msg_ = *input_joint_command_.readFromRT();
+
+  // if message exists, load values into references
+  if (joint_command_msg_.get())
+  {
+    for (auto i = 0ul; i < joint_command_msg_->positions.size(); i++)
+    {
+      position_reference_[i].get() = joint_command_msg_->positions[i];
+    }
+    for (auto i = 0ul; i < joint_command_msg_->velocities.size(); i++)
+    {
+      velocity_reference_[i].get() = joint_command_msg_->velocities[i];
+    }
+  }
+
+  return controller_interface::return_type::OK;
+}
+
+controller_interface::return_type AdmittanceController::update_and_write_commands(
+  const rclcpp::Time & time, const rclcpp::Duration & period)
+{
+  // Realtime constraints are required in this function
+  if (!admittance_)
+  {
+    return controller_interface::return_type::ERROR;
+  }
+
+  // update input reference from chainable interfaces
+  read_state_reference_interfaces(reference_);
+
+  // get all controller inputs
+  read_state_from_hardware(joint_state_, ft_values_);
+
+  // apply admittance control to reference to determine desired state
+  admittance_->update(joint_state_, ft_values_, reference_, period, reference_admittance_);
+
+  // write calculated values to joint interfaces
+  write_state_to_hardware(reference_admittance_);
+
+  // Publish controller state
+  state_publisher_->lock();
+  state_publisher_->msg_ = admittance_->get_controller_state();
+  state_publisher_->unlockAndPublish();
+
+  return controller_interface::return_type::OK;
+}
+
+controller_interface::CallbackReturn AdmittanceController::on_deactivate(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  force_torque_sensor_->release_interfaces();
+
+  return LifecycleNodeInterface::on_deactivate(previous_state);
+}
+
+controller_interface::CallbackReturn AdmittanceController::on_cleanup(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+controller_interface::CallbackReturn AdmittanceController::on_error(
+  const rclcpp_lifecycle::State & previous_state)
+{
+  if (!admittance_)
+  {
+    return controller_interface::CallbackReturn::ERROR;
+  }
+  admittance_->reset(num_joints_);
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+void AdmittanceController::read_state_from_hardware(
+  trajectory_msgs::msg::JointTrajectoryPoint & state_current,
+  geometry_msgs::msg::Wrench & ft_values)
+{
+  // Fill fields of state_reference argument from hardware state interfaces. If the hardware does
+  // not exist or the values are nan, the corresponding state field will be set to empty.
+
+  // if any interface has nan values, assume state_reference is the last command state
+  for (auto joint_ind = 0ul; joint_ind < num_joints_; joint_ind++)
+  {
+    for (auto inter_ind = 0ul; inter_ind < 3; inter_ind++)
```

Comment:
Please use the approach from JTC, which uses explicit flags for the interfaces.

Reply:
I don't understand the comment, so I will try to explain this method and hopefully it will help. `position_reference_` and `velocity_reference_` point to the data that will be written to by other controllers via claiming the command interface. This method transfers that data to the `state_reference` which is used in the update loop. If nan values are written to the reference, then the last valid reference will be used.